compiler/rustc_trait_selection/src/solve/assembly/structural_traits.rs

```diff
@@ -77,10 +77,22 @@ pub(in crate::solve) fn instantiate_constituent_tys_for_auto_trait<'tcx>(
         ty::Adt(def, substs) => Ok(def.all_fields().map(|f| f.ty(tcx, substs)).collect()),
 
         ty::Alias(ty::Opaque, ty::AliasTy { def_id, substs, .. }) => {
-            // We can resolve the `impl Trait` to its concrete type,
-            // which enforces a DAG between the functions requiring
-            // the auto trait bounds in question.
-            Ok(vec![tcx.type_of(def_id).subst(tcx, substs)])
+            match ecx.solver_mode() {
```

Comment:
instead: if we're either in `Reveal::All` or can define the opaque, don't add builtin candidate via opaque type leakage. Go through actual normalize instead. This is strictly better than using `type_of` directly

Reply:
I moved this check out of `instantiate_constituent_tys_for_auto_trait` because I didn't want to have to pass the param-env in to check the reveal mode.